tt-stl AnyIterator not compatible with GCC15's libstdc++

### Describe the bug

It looks like tt-stl's AnyIterator defined in any_range.hpp is trying to call the deleted default constructor when used with GCC 15's library stack. This happens both when compiling GCC 15 and Clang 20. I've been running into issues with GCC 15 in general, so this might be on GCC's end. Investigating.

```
[3/227] Building CXX object tests/tt_metal/tt_metal/stl/CMakeFiles/unit_tests_stl_smoke.dir/test_any_range.cpp.o
FAILED: tests/tt_metal/tt_metal/stl/CMakeFiles/unit_tests_stl_smoke.dir/test_any_range.cpp.o 
/usr/lib64/ccache/clang++ -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -DFMT_HEADER_ONLY=1 -DTT_ENABLE_LIGHT_METAL_TRACE=1 -I./tt_metal/api -I./tt_metal/api/tt-metalium -I./tt_metal/third_party/umd/device/api -I./tt_metal/hostdevcommon/api -I./tt_stl/. -I./tt_metal -I./tt_metal/impl -I./tt_metal/impl/debug -I./tt_metal/impl/profiler -I./tt_metal/include -I. -I./tt_metal/common -I./tt_metal/third_party/umd/src/firmware/riscv -I./.cpmcache/simd-everywhere/b3b426f78574ef837b17f42e86bab88314c5e4db -I./tt_metal/llrt -I./tt_metal/hw/inc -I./tt_metal/hw -I./tt_metal/fabric -I./.cpmcache/flatbuffers/863d0e04925ab8956bc26ae6121c40fe49c9d992/include -I./tests -I./tests/tt_metal/tt_metal/common -I./tests/tt_metal/tt_metal/stl -I./tests/tt_metal/tt_metal/stl/common -isystem ./.cpmcache/magic_enum/4d76fe0a5b27a0e62d6c15976d02b33c54207096/include -isystem ./.cpmcache/fmt/69912fb6b71fcb1f7e5deca191a2bb4748c4e7b6/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/core/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/assert/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/config/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/static_assert/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/throw_exception/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/container/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/intrusive/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/move/include -isystem ./tt_metal/third_party/tracy/public -isystem ./.cpmcache/nlohmann_json/b88ca108e0b5a597e859a21658d11fa5f1feb410/include -isystem ./.cpmcache/reflect/5110b6ebd4e43b1aa43e01e8c713c100c2b80fda -isystem ./.cpmcache/googletest/96129d89f45386492ae46d6bb8c027bc3df5f949/googlemock/include -isystem ./.cpmcache/googletest/96129d89f45386492ae46d6bb8c027bc3df5f949/googlemock -isystem ./.cpmcache/googletest/96129d89f45386492ae46d6bb8c027bc3df5f949/googletest/include -isystem ./.cpmcache/googletest/96129d89f45386492ae46d6bb8c027bc3df5f949/googletest -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/algorithm/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/array/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/bind/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/concept_check/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/preprocessor/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/type_traits/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/exception/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/smart_ptr/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/tuple/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/function/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/iterator/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/detail/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/fusion/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/container_hash/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/describe/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/mp11/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/function_types/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/mpl/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/predef/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/utility/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/io/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/typeof/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/functional/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/optional/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/range/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/conversion/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/regex/include -isystem ./.cpmcache/boost/801d416ee6f595cdc380b47d35af37e26dae8fca/libs/unordered/include -O3 -g -DDEBUG -std=c++20 -fPIC -fsized-deallocation -mavx2 -fno-lto -fPIC -fvisibility-inlines-hidden -Wall -pipe -Wno-deprecated-declarations -Wno-bool-operation -Wno-pessimizing-move -Wno-range-loop-construct -Wno-unused-but-set-variable -Wno-unused-function -Wno-unused-variable -Wno-c++11-narrowing -Wno-deprecated-this-capture -Wno-deprecated-volatile -Wno-for-loop-analysis -Wno-logical-op-parentheses -Wno-mismatched-tags -Wno-missing-braces -Wno-reorder-ctor -Wno-self-assign-overloaded -Wno-self-assign -Wno-self-move -Wno-shift-op-parentheses -Wno-unused-lambda-capture -Wno-unused-local-typedef -Wno-unused-private-field -Wno-unknown-pragmas -Wno-int-to-pointer-cast -MD -MT tests/tt_metal/tt_metal/stl/CMakeFiles/unit_tests_stl_smoke.dir/test_any_range.cpp.o -MF tests/tt_metal/tt_metal/stl/CMakeFiles/unit_tests_stl_smoke.dir/test_any_range.cpp.o.d -o tests/tt_metal/tt_metal/stl/CMakeFiles/unit_tests_stl_smoke.dir/test_any_range.cpp.o -c ./tests/tt_metal/tt_metal/stl/test_any_range.cpp
In file included from ./tests/tt_metal/tt_metal/stl/test_any_range.cpp:5:
In file included from ./.cpmcache/googletest/96129d89f45386492ae46d6bb8c027bc3df5f949/googletest/include/gtest/gtest.h:54:
In file included from /usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/iomanip:47:
In file included from /usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/ios_base.h:43:
In file included from /usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/locale_classes.h:42:
In file included from /usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/string:50:
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_iterator.h:182:37: error: call to deleted constructor of 'tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>'
  182 |       _GLIBCXX_NOEXCEPT_IF(noexcept(_Iterator()))
      |                                     ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/type_traits:3557:46: note: in instantiation of exception specification for 'reverse_iterator' requested here
 3557 |   inline constexpr bool is_constructible_v = __is_constructible(_Tp, _Args...);
      |                                              ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/concepts:162:30: note: in instantiation of variable template specialization 'std::is_constructible_v<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>>' requested here
  162 |       = destructible<_Tp> && is_constructible_v<_Tp, _Args...>;
      |                              ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/concepts:162:30: note: while substituting template arguments into constraint expression here
  162 |       = destructible<_Tp> && is_constructible_v<_Tp, _Args...>;
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/concepts:166:37: note: while checking the satisfaction of concept 'constructible_from<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>>' requested here
  166 |     concept default_initializable = constructible_from<_Tp>
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/concepts:166:37: note: while substituting template arguments into constraint expression here
  166 |     concept default_initializable = constructible_from<_Tp>
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/concepts:282:44: note: (skipping 20 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  282 |     concept semiregular = copyable<_Tp> && default_initializable<_Tp>;
      |                                            ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_algo.h:1782:8: note: in instantiation of function template specialization 'std::move_backward<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>, std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>>' requested here
 1782 |               _GLIBCXX_MOVE_BACKWARD3(__first, __i, __i + 1);
      |               ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_algobase.h:887:53: note: expanded from macro '_GLIBCXX_MOVE_BACKWARD3'
  887 | #define _GLIBCXX_MOVE_BACKWARD3(_Tp, _Up, _Vp) std::move_backward(_Tp, _Up, _Vp)
      |                                                     ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_algo.h:1818:9: note: in instantiation of function template specialization 'std::__insertion_sort<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1818 |           std::__insertion_sort(__first, __first + int(_S_threshold), __comp);
      |                ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_algo.h:1909:9: note: in instantiation of function template specialization 'std::__final_insertion_sort<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1909 |           std::__final_insertion_sort(__first, __last, __comp);
      |                ^
/usr/lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/bits/stl_algo.h:4841:12: note: in instantiation of function template specialization 'std::__sort<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 4841 |       std::__sort(__first, __last, __gnu_cxx::__ops::__iter_less_iter());
      |            ^
./tests/tt_metal/tt_metal/stl/test_any_range.cpp:70:10: note: in instantiation of function template specialization 'std::sort<std::reverse_iterator<tt::stl::AnyIterator<int &, std::random_access_iterator_tag, 40>>>' requested here
   70 |     std::sort(range.rbegin(), range.rend());
      |          ^
./tt_stl/./tt_stl/any_range.hpp:364:5: note: 'AnyIterator' has been explicitly marked deleted here
  364 |     AnyIterator() = delete;
      |     ^
1 error generated.
```

### Steps to reproduce the issue

Build with GCC 15 or Clang 20.

### Expected behavior

Built.

### Please complete the following environment information

Fedora 42
```
$ gcc --version
gcc (GCC) 15.0.1 20250329 (Red Hat 15.0.1-0)
```
```
$ clang --version
clang version 20.1.2 (Fedora 20.1.2-3.fc42)
```

--- a/tt_stl/tt_stl/any_range.hpp
+++ b/tt_stl/tt_stl/any_range.hpp
@@ -361,8 +361,6 @@ private:
     }
 
 public:
-    AnyIterator() = delete;
-
     AnyIterator(const AnyIterator& other) { other.iterator_adaptor().uninitialized_copy_to(bytes); }
 
     AnyIterator(AnyIterator&& other) noexcept { other.iterator_adaptor().uninitialized_move_to(bytes); }
@@ -457,6 +455,9 @@ private:
         : std::enable_if<std::is_base_of_v<std::forward_iterator_tag, iterator_category>, enable_if_forward> {};
 
 public:
+    template <class TEnable = enable_if_forward, class = typename TEnable::type>
+    AnyIterator() noexcept : AnyIterator(static_cast<std::add_pointer_t<reference>>(nullptr)) {}
+
     template <class TEnable = enable_if_forward, class = typename TEnable::type>
     [[nodiscard]] AnyIterator operator++(int) {
         static_assert(std::is_same_v<TEnable, enable_if_forward>);

